src/backend/pipeline/cont_adhoc.c

```diff
@@ -0,0 +1,1000 @@
+
+	PortalStart(portal, NULL, 0, GetActiveSnapshot());
+
+	(void) PortalRun(portal,
+			FETCH_ALL,
+			true,
+			receiver,
+			receiver,
+			NULL);
+
+	(*receiver->rDestroy) (receiver);
+	PortalDrop(portal, false);
+	PopActiveSnapshot();
+}
+
+typedef struct CleanupData
+{
+	ContinuousViewData *cont_view_data;
+	AdhocWorkerState *worker_state;
+} CleanupData;
+
+/* To be called when adhoc query is finished */
+static void
+cleanup(int code, Datum arg)
+{
+	CleanupData *cleanup;
+	AbortCurrentTransaction();
+	
+	cleanup = (CleanupData *)(arg);
+
+	TupleBufferBatchReaderReset(cleanup->worker_state->reader);
+	TupleBufferCloseBatchReader(cleanup->worker_state->reader);
+
+	release_cont_query_proc();
+
+	StartTransactionCommand();
+	AdhocMgrCleanupContinuousView(cleanup->cont_view_data->view);
+	CommitTransactionCommand();
+}
+
+/* Initialize the worker, combiner and view modules, then loop until done.
+ * The current exit conditions are when the query is cancelled, or when the frontend
+ * goes away (the heartbeat will fail to send) */
+static void
+exec_adhoc_query(SelectStmt *stmt, const char *s)
+{
+	int numCols = 1;
+	AttrNumber one = 1;
+	AttrNumber *keyColIdx = &one;
+
+	AdhocWorkerState *worker_state = 0;
+	AdhocCombinerState *combiner_state = 0;
+	AdhocViewState *view_state = 0;
+	AdhocDestReceiver *adhoc_receiver = 0;
+	DestReceiver *worker_receiver = 0;
+
+	MemoryContext run_cxt = CurrentMemoryContext;
+	ContinuousViewData view_data;
+	Tuplestorestate *batch = 0;
+	CleanupData cleanup_data;
+
+	ResourceOwner owner = 
+			ResourceOwnerCreate(CurrentResourceOwner, "WorkerResourceOwner");
+
+	acquire_cont_query_proc();
+
+	if (!MyContQueryProc)
+	{
+		elog(ERROR, "too many adhoc processes");
+	}
+
+	StartTransactionCommand();
+	CurrentResourceOwner = owner;
+	MemoryContextSwitchTo(run_cxt);
+	view_data = init_cont_view(stmt, s);
+	CommitTransactionCommand();
+
+	StartTransactionCommand();
+	CurrentResourceOwner = owner;
+	MemoryContextSwitchTo(run_cxt);
+	worker_receiver = CreateDestReceiver(DestTuplestore);
+
+	batch = tuplestore_begin_heap(true, true, 
+				continuous_query_combiner_work_mem);
+
+	SetTuplestoreDestReceiverParams(worker_receiver, batch,
+			CurrentMemoryContext, true);
+
+	worker_state = init_adhoc_worker(view_data, worker_receiver);
+	combiner_state = init_adhoc_combiner(view_data, batch);
+
+	if (combiner_state->existing)
+	{
+		keyColIdx = combiner_state->existing->keyColIdx;
+		numCols = combiner_state->existing->numCols;
+	}
+
+	adhoc_receiver = CreateAdhocDestReceiver(combiner_state->is_agg,
+										 keyColIdx,
+										 numCols);
+
+	view_state = init_adhoc_view(view_data, combiner_state->result,
+				 (DestReceiver *) adhoc_receiver);
+	view_state->slot = combiner_state->slot;
+
+	CommitTransactionCommand();
+
+	cleanup_data.cont_view_data = &view_data;
+	cleanup_data.worker_state = worker_state;
+
+	PG_ENSURE_ERROR_CLEANUP(cleanup, PointerGetDatum(&cleanup_data));
+	{
+		while (true)
```

Comment:
Before getting into this loop we should set `XactReadOnly = true` since we don't perform any writes, but reset it to true afterwards so we can do the clean up.